### PR TITLE
[async] Remove ClearListStmt from possibly fused serial tasks

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #

I don't think this will actually be effective at HEAD yet, because currrently we do listgen optimization before fusion. However, as I added another fusion pass before the listgen optimization (locally), it did seem to work as expected? (Dots were generated from this new test case)

1. initial

Screen Shot 2020-09-23 at 22 22 02

2. fuse

Screen Shot 2020-09-23 at 22 26 35

3. listgen optimize

Screen Shot 2020-09-23 at 22 26 02

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Review by yuanming-hu — APPROVED** [review]
Great! Thank you so much!

Since `listgen_opt` now works after `fusion`, maybe we can change

https://github.com/taichi-dev/taichi/blob/8c5169f8556bed328544eee628665859b51afc47/taichi/program/async_engine.cpp#L452-L483

into

```
  while (modified) {
    modified = true;
    while (modified) {
      modified = false;
      if (program->config.async_opt_listgen)
        while (sfg->optimize_listgen()) {
          debug_sfg("listgen");
          modified = true;
        }
      sfg->verify();
      if (program->config.async_opt_dse)
        while (sfg->optimize_dead_store()) {
          debug_sfg("dse");
          modified = true;
        }
      sfg->verify();
      if (program->config.async_opt_activation_demotion)
        while (sfg->demote_activation()) {
          debug_sfg("act");
          modified = true;
        }
      sfg->verify();
      if (program->config.async_opt_fusion)
        while (sfg->fuse()) {
          debug_sfg("fuse");
          modified = true;
        }
      sfg->verify();
    }
  }
```

so that all optimizations can be applied iteratively? :-)

**Review by xumingkuan — APPROVED** [review]
Didn't get the point why the serial task in the third graph has no output edges... The code itself LGTM.

File: tests/python/test_sfg.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+@ti.test(require=ti.extension.async_mode, async_mode=True)
+def test_remove_clear_list_from_fused_serial():
+    x = ti.field(ti.i32)
+    y = ti.field(ti.i32)
+    z = ti.field(ti.i32, shape=())
+
+    n = 32
+    ti.root.bitmasked(ti.i, n).place(x)
+    ti.root.bitmasked(ti.i, n).place(y)
```
[the remark this answers is not in the document]
Comment (k-ye) [reply tests/python/test_sfg.py]: Ah yes, i switched to `root.pointer.place` now :) I know it's a bit pointless to directly put a `place` after `pointer`, but it makes it a bit simpler to assert which indices are activated...

**Comment by k-ye** [comment]
> Since listgen_opt now works after fusion, maybe we can change

Ack. Let me do this in another PR. In that one, I'd also like to cleanup `debug_sfg` a bit, so that it has a monotonicallying increasing **sync ID**. E.g.

foo_sync000_initial
foo_sync000_fuse_000
foo_sync000_listgen_000
...
foo_sync000_fuse_001
foo_sync000_listgen_001
...
foo_sync000_final

foo_sync001_initial
...

WDYT?

> Didn't get the point why the serial task in the third graph has no output edges... 

I think step 3's listgen has reverted the fused serial task to the original `serial_z` task, which looked the same as the one in the first graph..

**Comment by xumingkuan** [comment]
> > Since listgen_opt now works after fusion, maybe we can change
> 
> Ack. Let me do this in another PR. In that one, I'd also like to cleanup `debug_sfg` a bit, so that it has a monotonicallying increasing **sync ID**. E.g.
> 
> foo_sync000_initial
> foo_sync000_fuse_000
> foo_sync000_listgen_000
> ...
> foo_sync000_fuse_001
> foo_sync000_listgen_001
> ...
> foo_sync000_final
> 
> foo_sync001_initial
> ...
> 
> WDYT?

Agreed! When a program which syncs many times runs 2 times with different numbers of iterations of optimization, the file name could be confusing otherwise.

> > Didn't get the point why the serial task in the third graph has no output edges...
> 
> I think step 3's listgen has reverted the fused serial task to the original `serial_z` task, which looked the same as the one in the first graph..

Cool!

**Pull request merged**